File: js/Plane.js
import {Vector3} from "./Vector3.js";

export class Plane {

    // Plane is defined by an origin point R, and a normal vector N
    // point P is on the plane if and only if RP.N = 0
    constructor(origin, normal) {
        this.origin = origin;
        this.normal = Vector3.normalize(normal);
    }

    // Plane across 2 points
    static across(p1, p2) {
        const normal = new Vector3(p2.x - p1.x, p2.y - p1.y, p2.z - p1.z);
        const middle = new Vector3((p1.x + p2.x) / 2, (p1.y + p2.y) / 2, (p1.z + p2.z) / 2);
        return new Plane(middle, normal);
    }

    // Plane by 2 points on xy orthogonal to z
    static by(p1, p2) {
        // Turn 90° on the right (x,y) to (y,-x)
        const normal = new Vector3((p2.y - p1.y), -(p2.x - p1.x), 0);
        return new Plane(p1, normal);
    }

    // Plane orthogonal to a segment [p1, p2] passing by point
    static orthogonal(p1, p2, point) {
        const normal = new Vector3(p2.x - p1.x, p2.y - p1.y, p2.z - p1.z);
        return new Plane(point, normal);
    }

    // Plane through point containing direction [p1, p2] (crease parallel to the segment).
    // Prefer a vertical plane (normal ⊥ Z), matching Plane.by when the paper is flat.
    static parallel(p1, p2, point) {
        const dx = p2.x - p1.x, dy = p2.y - p1.y;
        // direction × (0,0,1) = (dy, -dx, 0)
        if (Math.abs(dx) > 1e-12 || Math.abs(dy) > 1e-12) {
            return new Plane(point, new Vector3(dy, -dx, 0));
        }
        // Segment projects to a point in xy (along Z): any vertical plane through point
        return new Plane(point, new Vector3(1, 0, 0));
    }

}
// 32

File: src/plane.rs
use crate::vec3::Vec3;
use crate::point::Point;

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Plane {
    pub origin: Vec3,
    pub normal: Vec3,
}

impl Plane {
    // Plane is defined by an origin point R, and a normal vector N
    pub fn new(origin: Vec3, normal: Vec3) -> Self {
        Plane {
            origin,
            normal: Vec3::normalize(normal),
        }
    }

    // Plane across 2 points
    pub fn across(p1: Vec3, p2: Vec3) -> Self {
        let normal = Vec3::sub(p2, p1);
        let middle = Vec3::scale(Vec3::add(p1, p2), 0.5);
        Plane::new(middle, normal)
    }

    // Plane by 2 points on xy orthogonal to z
    pub fn by(p1: Vec3, p2: Vec3) -> Self {
        // Turn 90° on the right (x,y) to (y,-x)
        let normal = Vec3::new(p2.y - p1.y, -(p2.x - p1.x), 0.0);
        Plane::new(p1, normal)
    }

    // Plane orthogonal to a segment [p1, p2] passing by point
    pub fn orthogonal(p1: Vec3, p2: Vec3, point: Vec3) -> Self {
        let normal = Vec3::sub(p2, p1);
        Plane::new(point, normal)
    }

    pub fn plane_to_point_signed_distance(&self, p: &Point) -> f32 {
        let v = Vec3::sub(Vec3::from_point(p), self.origin);
        Vec3::dot(v, self.normal)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_across() {
        let p1 = Vec3::new(10.0, 50.0, 0.0);
        let p2 = Vec3::new(30.0, 50.0, 0.0);
        let plane = Plane::across(p1, p2);
        assert_eq!(plane.normal.x, 1.0);
        assert_eq!(plane.normal.y, 0.0);
        assert_eq!(plane.normal.z, 0.0);
        assert_eq!(Vec3::dot(plane.origin, plane.normal), 20.0);
    }

    #[test]
    fn test_by() {
        let p1 = Vec3::new(30.0, 50.0, 0.0);
        let p2 = Vec3::new(10.0, 50.0, 0.0);
        let plane = Plane::by(p1, p2);
        assert_eq!(plane.normal.x, 0.0);
        assert_eq!(plane.normal.y, 1.0);
        assert_eq!(plane.normal.z, 0.0);
        assert_eq!(Vec3::dot(plane.origin, plane.normal), 50.0);
    }

    #[test]
    fn test_orthogonal() {
        let p1 = Vec3::new(10.0, 50.0, 0.0);
        let p2 = Vec3::new(30.0, 50.0, 0.0);
        let pt = Vec3::new(20.0, 0.0, 0.0);
        let plane = Plane::orthogonal(p1, p2, pt);
        assert_eq!(plane.normal.x, 1.0);
        assert_eq!(plane.normal.y, 0.0);
        assert_eq!(plane.normal.z, 0.0);
        assert_eq!(Vec3::dot(plane.origin, plane.normal), 20.0);
    }
}
